File: src/P014/p014.go
package problem014

import "strconv"

func Solve() string {
	var chainLengths [1000000]int
	chainLengths[1] = 1

	maxChainLength := 0
	maxChainLengthStart := 0

	for start := 2; start < 1000000; start++ {
		n := start
		length := 0
		for n >= start {
			length++
			if n%2 == 0 {
				n /= 2
			} else {
				n = 3*n + 1
			}
		}
		chainLengths[start] = length + chainLengths[n]
		if chainLengths[start] > maxChainLength {
			maxChainLength = chainLengths[start]
			maxChainLengthStart = start
		}
	}
	return strconv.Itoa(maxChainLengthStart)
}

File: src/P014/p014.ts
const getCollatzSequenceLength = (
  start: number,
  chainLengths: Record<number, number>
): number => {
  let n = start;
  let len = 0;
  while (chainLengths[n] === undefined) {
    len++;
    n = n % 2 === 0 ? n / 2 : 3 * n + 1;
  }
  chainLengths[start] = len + chainLengths[n];
  return chainLengths[start];
};

export const solve = (): string => {
  const chainLengths: Record<number, number> = {};
  chainLengths[1] = 1;

  let maxChainLength = 0;
  let maxChainLengthStart = 0;
  for (let i = 1; i < 1000000; i++) {
    const len = getCollatzSequenceLength(i, chainLengths);
    if (len > maxChainLength) {
      maxChainLength = len;
      maxChainLengthStart = i;
    }
  }
  return maxChainLengthStart.toString();
};
